src/main/proto/com/google/api/codegen/config.proto

```diff
@@ -316,62 +318,77 @@ message CodeSampleValues {
-  //    several placeholders.
+  // * `DOTTED.RESOURCE.NAME%FIELD=VALUE`, if `NAME` is required to be
+  //   a specially formatted string using `FIELD` as one of possibly
+  //   several placeholders.
   repeated string parameters = 4;
 }
 
-// `SampleConfiguration` defines combinations of "molds" (sample code
-// structures as output by the client library generator) and
-// `CodeSampleValues` that should be combined to generate samples.
+// `SampleConfiguration` defines combinations of "calling forms"
+// (sample code structures as output by the client library generator)
+// and sample value sets that should be combined to generate samples.
 message SampleConfiguration {
-  // `Value` references one or more `CodeSampleValues` and provides a
-  // hint as to where it should be presented.
-  message Value {
-    // One or more of the `CodeSampleValues.id`s defined for this
-    // method. If more than one is specified and `explorer_sample` is
-    // set, the hint for relative order of the samples generated is
-    // taken to be the order of this list. If not specified, `ids`
-    // defaults to a list of all defined values for
-    // `MethodConfigProto.code_sample_values.id`. If any element of
-    // `ids` is invalid (i.e., is not found in
-    // `MethodConfigProto.code_sample_values.id`) and `delete` is not
-    // set, then an fatal error is raised at code-generation time.
-    repeated string ids = 1;
-
-    // Whether the samples specified by this configuration should also
-    // be used as configurable "Explorer Samples" suitable for
-    // inclusion in a UI element.
-    bool explorer_sample = 2;
-
-    // Whether the samples specified by this configuration should also
-    // be included within code code comments in the surface of each
-    // generated client library.
-    bool in_code_sample = 3;
-
-    // A hint as to the user-visible position of this
-    // `SampleConfiguration` relative to other
-    // `SampleConfiguration`s. This is only used when generating
-    // "Explorer Samples", which only happens if `explorer_sample` is
-    // set.
-    uint32 position = 4;
-  }
 
-  // The sample code mold (defined in the client library generator for
-  // various method types) to which this `SampleConfiguration` applies
-  //
-  // TODO(vchudnov-g): Point to a reference listing the possible molds
-  // for each method type.
-  repeated string molds = 1;
+  // `SampleTypeConfiguration` defines the samples that will show up
+  // as a particular sample type. If order matters for this sample
+  // type, the order of the samples is taken to be primarily the order
+  // in which the `calling_forms` appear, and secondarily, for each
+  // such form, the order in which the `value_sets` appear.
+  message SampleTypeConfiguration {
+    // One or more expressions matching the method calling forms
+    // (defined by the client generator) applicable to this
+    // method. Any expressions that do not match at least one calling
```

Comment:
> Any expressions that do not match at least one calling form cause a fatal error during code generation

Could you clarify this?

In the test configuration, I see `request_based` and `callable` being used as expressions here. I assume this refers to the distinction in Java between request-object methods and flattened method. (Is that right?) If so, `request_based` doesn't apply to most of the dynamic languages, which have no request-object methods. I believe the generators for those languages should simply ignore configuration for that calling form, but not raise an error.

Reply:
That's a good point. i guess I was thinking forms "that do not match at least one calling form **in at least one language**" will cause an error, but otherwise they will be ignored by the languages that don't employ that calling form. My motivation was to catch configuration errors as early and prominently as possible; a situation like this is an obvious error.

However, now that you bring it up, I wonder whether this really confuses the semantics. Maybe it's better to say all unrecognized expressions are ignored, regardless of whether they exist for any language. I suppose we could issue an INFO message for each such unrecognized instance (though that would potentially be a lot of instances and I don't know whether people do/would look at those messages). Thoughts?

[Separately, I wonder whether eventually we'll want to expand `SampleTypeConfiguration` to include a `language` field. It's premature to do it now, though.]

Reply:
> Maybe it's better to say all unrecognized expressions are ignored, regardless of whether they exist for any language.

I prefer this because it allows a (future, modular) world where a generator for a particular language doesn't need to know about the behavior other languages' generators. An INFO message for unrecognized expressions might still be helpful for debugging, as long as the tooling doesn't halt.